tgui/packages/tgui/interfaces/Fabricator.tsx

```diff
@@ -0,0 +1,404 @@
+import { useBackend, useSharedState } from '../backend';
+import { Stack, Section, Button, Input, Icon, Tabs, Dimmer } from '../components';
+import { Window } from '../layouts';
+import { Material, MaterialAmount, MaterialFormatting, Materials, MATERIAL_KEYS } from './common/Materials';
+import { Fragment } from 'inferno';
+import { sortBy } from 'common/collections';
+
+type MaterialMap = Partial<Record<keyof typeof MATERIAL_KEYS, number>>;
+
+/**
+ * A single design that the fabricator can print.
+ */
+type Design = {
+  /**
+   * The name of the design.
+   */
+  name: string;
+
+  /**
+   * A human-readable description of the design.
+   */
+  desc: string;
+
+  /**
+   * The individual material cost to print the design, adjusted for the
+   * fabricator's part efficiency.
+   */
+  cost: MaterialMap;
+
+  /**
+   * A reference to the design's design datum.
+   */
+  id: string;
+
+  /**
+   * The categories the design should be present in.
+   */
+  categories?: string[];
+};
+
+type FabricatorData = {
+  /**
+   * The materials available to the fabricator, via ore silo or local storage.
+   */
+  materials: Material[];
+
+  /**
+   * The name of the fabricator, as displayed on the title bar.
+   */
+  fab_name: string;
+
+  /**
+   * Whether mineral access is disabled from the ore silo (contact the
+   * quartermaster).
+   */
+  on_hold: boolean;
+
+  /**
+   * The set of designs that this fabricator can print, ordered by their ID.
+   */
+  designs: Record<string, Design>;
+
+  /**
+   * Whether the fabricator is currently printing an item.
+   */
+  busy: boolean;
+};
+
+/**
+ * Categories present in this object are not rendered to the final fabricator
+ * UI.
+ */
+const BLACKLISTED_CATEGORIES: Record<string, boolean> = {
+  'initial': true,
+  'core': true,
+  'hacked': true,
+};
+
+/**
+ * A dummy category that, when selected, renders ALL recipes to the UI.
+ */
+const ALL_CATEGORY = '__ALL';
+
+const SearchBar = (
+  props: { searchText: string; setSearchText: (text: string) => void },
```

Comment:
It's usually more readable to extract prop types to a separate type alias, e.g. `type SearchBarProps`.